File: server/apps/nexotype/schemas/asset/oligonucleotide_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class OligonucleotideCreate(BaseModel):
    """Schema for creating a new oligonucleotide"""
    uid: str
    name: str
    project_code: str | None = None
    sequence_na: str
    modification_type: str | None = None


class OligonucleotideUpdate(BaseModel):
    """Schema for updating an oligonucleotide"""
    uid: str = None
    name: str = None
    project_code: str | None = None
    sequence_na: str = None
    modification_type: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class OligonucleotideDetail(BaseModel):
    """Schema for oligonucleotide details"""
    id: int
    uid: str
    name: str
    project_code: str | None = None
    asset_type: str
    sequence_na: str
    modification_type: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class OligonucleotideListResponse(BaseModel):
    """Response schema for listing oligonucleotides"""
    success: bool
    data: list[OligonucleotideDetail] | None = None
    count: int = 0
    error: str | None = None


class OligonucleotideResponse(BaseModel):
    """Response schema for single oligonucleotide operations"""
    success: bool
    data: OligonucleotideDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/asset/oligonucleotide.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const OligonucleotideSchema = z.object({
  id: z.number(),
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  asset_type: z.string().max(50),
  sequence_na: z.string().min(1),
  modification_type: z.string().max(50).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateOligonucleotideSchema = z.object({
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  sequence_na: z.string().min(1),
  modification_type: z.string().max(50).nullable().optional(),
});

export const UpdateOligonucleotideSchema = z.object({
  uid: z.string().min(1).max(100).optional(),
  name: z.string().min(1).max(255).optional(),
  project_code: z.string().max(50).nullable().optional(),
  sequence_na: z.string().min(1).optional(),
  modification_type: z.string().max(50).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type Oligonucleotide = z.infer<typeof OligonucleotideSchema>;
export type CreateOligonucleotide = z.infer<typeof CreateOligonucleotideSchema>;
export type UpdateOligonucleotide = z.infer<typeof UpdateOligonucleotideSchema>;

// ==========================================
// Response Types
// ==========================================

export type OligonucleotideResponse = {
  success: boolean;
  data?: Oligonucleotide;
  error?: string;
};

export type OligonucleotidesResponse = {
  success: boolean;
  data?: Oligonucleotide[];
  count?: number;
  error?: string;
};
